model-optimizer/extensions/back/TransposeDFT.py

```diff
@@ -0,0 +1,34 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+
+from mo.back.replacement import BackReplacementPattern
+from mo.graph.graph import Graph
+
+
+class TransposeDFT(BackReplacementPattern):
+    """
+    In TF models, operation (I)FFTxD has some input shape, [N_0, ..., N_{r - 1}].
+
+    After the transformation SSliceComplexRolledFFTPackBlockReplacement, we have an input shape [N_0, ..., N_{r - 1}, 2]
+    for operation DFT or IDFT.
+
+    If the input rank in the TF model was greater than 2, we have [N_0, 2, N_1, ..., N_{r - 1}] as the input shape of
+    (I)DFT after the layout conversion, if the option '--disable_nhwc_to_nchw' is not specified.
+
+    But, generally speaking, according to DFT and IDFT specifications, the input shape [N_0, 2, N_1, ..., N_{r - 1}]
+    is not correct input shape for DFT and IDFT. Hence, we need to insert Transpose operations before and after (I)DFT
+    in such cases.
+
+    This transformation inserts such Transpose nodes, when the source model was the TF model, (I)DFT node has the
+    attribute 'need_insert_transposes_for_dft', and this attribute is True.
+    """
+    enabled = True
+    force_shape_inference = True
+    graph_condition = [lambda graph: graph.graph['fw'] == 'tf']
```

Comment:
`and graph.graph['layout'] != 'NCHW'`. Because if the original model is in NCHW layout you don't need to insert transposes.

Reply:
But this transformation is in the **back** stage. Hence, we have `graph.graph['layout'] == 'NCHW'` for any layout of the original model. Moreover, this transformation is executed only for DFT/IDFT with `need_insert_transposes_for_dft==True`.
Finally, the (I)DFT node has the attribute `need_insert_transposes_for_dft` with the value `True` only when the expression `not graph.graph['cmd_params'].disable_nhwc_to_nchw or graph.graph['layout'] == 'NHWC'` is `True` in the **front** stage (see line 37 in the file `model-optimizer/extensions/front/tf/TFFFTToDFT.py`).